src/Features/LanguageServer/Protocol/Handler/InlayHint/InlayHintHandler.cs

```diff
@@ -0,0 +1,134 @@
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler.InlayHint
+{
+    [ExportCSharpVisualBasicStatelessLspService(typeof(InlayHintHandler)), Shared]
+    [Method(Methods.TextDocumentInlayHintName)]
+    internal sealed class InlayHintHandler : ILspServiceDocumentRequestHandler<InlayHintParams, LSP.InlayHint[]?>
+    {
+        private readonly IGlobalOptionService _optionsService;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public InlayHintHandler(IGlobalOptionService optionsService)
+        {
+            _optionsService = optionsService;
+        }
+
+        public bool MutatesSolutionState => false;
+
+        public bool RequiresLSPSolution => true;
+
+        public TextDocumentIdentifier GetTextDocumentIdentifier(InlayHintParams request)
+            => request.TextDocument;
+
+        public async Task<LSP.InlayHint[]?> HandleRequestAsync(InlayHintParams request, RequestContext context, CancellationToken cancellationToken)
+        {
+            var document = context.GetRequiredDocument();
+            var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            var textSpan = ProtocolConversions.RangeToTextSpan(request.Range, text);
+
+            var inlineHintService = document.GetRequiredLanguageService<IInlineHintsService>();
+            var options = _optionsService.GetInlineHintsOptions(document.Project.Language);
+
+            var hints = await inlineHintService.GetInlineHintsAsync(document, textSpan, options, displayAllOverride: false, cancellationToken).ConfigureAwait(false);
+
+            using var _ = ArrayBuilder<LSP.InlayHint>.GetInstance(hints.Length, out var inlayHints);
+            var syntaxVersion = await document.GetSyntaxVersionAsync(cancellationToken).ConfigureAwait(false);
+            var inlayHintCache = context.GetRequiredLspService<InlayHintCache>();
+
+            // Store the members in the resolve cache so that when we get a resolve request for a particular
+            // member we can re-use the inline hint.
+            var resultId = inlayHintCache.UpdateCache(new InlayHintCache.InlayHintCacheEntry(hints, request.TextDocument, syntaxVersion));
+
+            for (var i = 0; i < hints.Length; i++)
+            {
+                var hint = hints[i];
+                var (label, leftPadding, rightPadding) = Trim(hint.DisplayParts);
+                var linePosition = text.Lines.GetLinePosition(hint.Span.Start);
+                var kind = hint.Ranking == InlineHintsConstants.ParameterRanking
+                    ? InlayHintKind.Parameter
+                    : InlayHintKind.Type;
+
+                // TextChange is calculated at the same time as the InlineHint,
+                // so it should not need to be resolved.
+                /*TextEdit[]? textEdits = null;
```

Comment:
Would remove this until the client side bug preventing us from providing them on resolve is fixed